File: go/all_paths_from_source_to_target.go
package leetcode

func allPathsSourceTarget(graph [][]int) [][]int {
	res := make([][]int, 0, 10)
	dfs(graph, 0, []int{0}, &res)
	return res
}

func dfs(graph [][]int, node int, partialRes []int, res *[][]int) {
	if node == len(graph)-1 {
		cpPartialRes := make([]int, len(partialRes))
		copy(cpPartialRes, partialRes)
		*res = append(*res, cpPartialRes)
		return
	}

	for _, nei := range graph[node] {
		partialRes = append(partialRes, nei)
		dfs(graph, nei, partialRes, res)
		partialRes = partialRes[:len(partialRes)-1]
	}
}

File: python/all-paths-from-source-to-target.py
from typing import List

class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        def dfs(node, partial_res, res):
            if node == len(graph)-1:
                res.append(partial_res[:])
                return
            
            for nei in graph[node]:
                partial_res.append(nei)
                dfs(nei, partial_res, res)
                partial_res.pop()
        
        res = []
        dfs(0, [0], res)
        return res